File: Algorithm/Merge Sort.java
public void mergeSort(int l, int r)
{
    if (l + 1 == r) return ;

    int mid = (l + r) / 2;
    mergeSort(l, mid);
    mergeSort(mid, r);

    int size = 0;
    int[] b = new int[r - l];
    for (int i = l, j = mid; ; ) {
        if (a[i] < a[j]) b[size++] = a[i++];
        else b[size++] = a[j++];

        if (i == mid) {
            while (j < r) b[size++] = a[j++];
            break;
        }
        if (j == r) {
            while (i < mid) b[size++] = a[i++];
            break;
        }
    }
    for (int i = 0; i < size; i++) a[l + i] = b[i];
}

File: Templates/merge-sort.cc
void merge_sort(int *a, int l, int r) // [l, r)
{
    if (l + 1 == r) return ;

    int mid = (l + r) / 2;
    merge_sort(a, l, mid);
    merge_sort(a, mid, r);

    int *b = new int[r - l];
    int p = l, q = mid, cnt = 0;
    while (p < mid && q < r) {
        if (a[p] < a[q]) b[cnt++] = a[p++];
        else b[cnt++] = a[q++];
    }
    while (p < mid) b[cnt++] = a[p++];
    while (q < r)   b[cnt++] = a[q++];

    for (int i = 0; i < cnt; i++) a[l + i] = b[i];

    delete[] b;
}
